src/goto-symex/symex_main.cpp

```diff
@@ -194,37 +213,62 @@ void goto_symext::operator()(
   const goto_programt::const_targett first,
   const goto_programt::const_targett limit)
 {
-  symex_entry_point(state, goto_functions, first, limit);
+  operator()(
+    state, get_function_from_goto_functions(goto_functions), first, limit);
+}
+
+void goto_symext::operator()(
+  statet &state,
+  const get_goto_functiont &get_goto_function,
+  const goto_programt::const_targett first,
+  const goto_programt::const_targett limit)
+{
+  symex_entry_point(state, get_goto_function, first, limit);
   while(state.source.pc->function!=limit->function || state.source.pc!=limit)
-    symex_threaded_step(state, goto_functions);
+    symex_threaded_step(state, get_goto_function);
+}
+
+void goto_symext::operator()(
+  const get_goto_functiont &get_goto_function,
+  const goto_programt &goto_program)
+{
+  statet state;
+  operator() (state, get_goto_function, goto_program);
 }
 
 /// symex starting from given program
 void goto_symext::operator()(
   const goto_functionst &goto_functions,
   const goto_programt &goto_program)
 {
-  statet state;
-  operator() (state, goto_functions, goto_program);
+  operator()(get_function_from_goto_functions(goto_functions), goto_program);
 }
 
 /// symex from entry point
-void goto_symext::operator()(const goto_functionst &goto_functions)
+void goto_symext::operator()(const get_goto_functiont &get_goto_function)
 {
-  goto_functionst::function_mapt::const_iterator it=
-    goto_functions.function_map.find(goto_functionst::entry_point());
-
-  if(it==goto_functions.function_map.end())
+  const goto_functionst::goto_functiont *start_function;
+  try
+  {
+    start_function = &get_goto_function(goto_functionst::entry_point());
+  }
+  catch(const std::out_of_range &error)
+  {
     throw "the program has no entry point";
+  }
```

Comment:
I'd suggest to turn this into a `PRECONDITION` - the problem can be detected much earlier and should detected by the caller. (I know this isn't really your bug, but since lots of this code is being touched here ... if you don't mind?)